src/sentry/ratelimits/cardinality.py

```diff
@@ -0,0 +1,139 @@
+"""
+A kind of limiter that limits set cardinality instead of a rate/count.
+
+The high-level concepts are very similar to `sentry.ratelimits.sliding_windows`.
+
+Instead of passing in numbers and getting back smaller numbers, however, the
+user passes in a set and gets back a smaller set. Set elements that have
+already been observed in any quota's window are "for free" and will not
+count towards any quota.
+
+The implementation hasn't been finalized yet, but we expect that under the hood
+this cardinality limiter will be more expensive to operate than a simple rate
+limiter, as it needs to keep track of already-seen set elements. The memory
+usage in Redis will most likely be proportional to the set size.
+
+This kind of limiter does not support prefix overrides, which practically means
+that there can only be a per-org or a global limit, not both at once.
+"""
+
+
+import time
+from dataclasses import dataclass
+from typing import Any, Collection, Iterator, Optional, Protocol, Sequence, Tuple, TypeVar
+
+from sentry.utils.services import Service
+
+Hash = str
+Timestamp = int
+
+
+@dataclass(frozen=True)
+class Quota:
+    # The number of seconds to apply the limit to.
+    window_seconds: int
+
+    # A number between 1 and `window_seconds`. Since `window_seconds` is a
+    # sliding window, configure what the granularity of that window is.
+    #
+    # If this is equal to `window_seconds`, the quota resets to 0 every
+    # `window_seconds`.  If this is a very small number, the window slides
+    # "more smoothly" at the expense of having much more redis keys.
+    #
+    # The number of redis keys required to enforce a quota is `window_seconds /
+    # granularity_seconds`.
+    granularity_seconds: int
+
+    #: How many units are allowed within the given window.
+    limit: int
+
+    def __post__init__(self) -> None:
+        assert self.window_seconds % self.granularity_seconds == 0
+
+
+@dataclass(frozen=True)
```

Comment:
I cannot remember if we had this discussion about the previous rate limiter implementation.
I think NamedTuples may have a lower performance impact than data classes (depends a lot on the python version) at the throughput we are talking about.

Reply:
yeah we had that conversation after introducing the Metadata class which significantly decreased performance in the indexer.

We can use namedtuple here and afaik mypy can type it. Should we change it everywhere then?

Reply:
Let's start with this PR. Then, I think, making the change in the other indexer classes will be a good idea.